src/inmanta/protocol/common.py

```diff
@@ -1158,46 +1192,215 @@ def wait(self, timeout: int = 60) -> None:
+        without wrapping in a `Result` object. If there are pages, simply chains results from multiple pages after each other.
+        """
+        page: "PageableResult[V]"
+        async for page in self._pages():
+            unwrapped: Sequence[V] = page.value()
+            for item in unwrapped:
+                yield item
+
+    async def _pages(self) -> AsyncIterator["PageableResult[V]"]:
         """
-        Helper method to iterate over all individual items in this result object.
-        This method will start at the first page and follow paging links.
+        Returns an async iterator over pages, starting from this one.
         """
+        result: Optional["PageableResult[V]"] = self
+        while result is not None:
+            yield result
+            result = await result.next_page()
 
-        if self._method_properties is None or self._client is None:
+    async def next_page(self) -> Optional["PageableResult[V]"]:
+        if self._client is None:
             raise Exception(
-                "The all() method cannot be called on this Result object. Make sure you "
+                "The next_page() method cannot be called on this Result object. Make sure you "
                 "set the client and method_properties parameters when constructing "
                 "a Result object manually (e.g. outside of a regular API call)."
             )
 
-        result = self
-        while result.code == 200:
-            if not result.result:
-                return
+        if not self.result:
+            return None
 
-            page = result.result.get(self._method_properties.envelope_key, [])
-            for item in page:
-                yield item
+        next_link_url = self.result.get("links", {}).get("next")
+        if not next_link_url:
+            return None
 
-            next_link_url = result.result.get("links", {}).get("next")
+        server_url = self._client._get_client_config()
+        url = server_url + next_link_url
+        headers = {"X-Inmanta-tid": self._environment} if self._environment else None
+        request = HTTPRequest(url=url, method="GET", headers=headers)
+        result: Result[list[V]] = self._client._decode_response(
+            await self._client.client.fetch(request), self.method_properties, self._environment
+        )
 
-            if not next_link_url:
-                return
+        return result.pageable()
 
-            server_url = self._client._get_client_config()
-            url = server_url + next_link_url
-            headers = {"X-Inmanta-tid": self._environment} if self._environment else None
-            request = HTTPRequest(url=url, method="GET", headers=headers)
-            result = self._client._decode_response(
-                await self._client.client.fetch(request), self._method_properties, self._environment
-            )
+
+class ClientCall(Awaitable[Result[R]]):
```

Comment:
Calling a method on a client now returns this object, instead of a `Coroutine[..., ..., Result]`. See `inmanta.protocol.endpoints.Client`.